# Part 1: Dataset Inspection

## Objective

The objective of this stage is to understand the structure, quality, and contents of the Telco Customer Churn dataset before performing any data cleaning or analysis.

## 1.1 Import Required Libraries

The required Python libraries are imported to enable data loading, manipulation, and analysis throughout the project.

In [3]:
import pandas as pd
import numpy as np

print("Setup successful!")

Setup successful!


## 1.2 Load the Dataset

The Telco Customer Churn dataset is loaded into a Pandas DataFrame to begin the data inspection process.

In [4]:
df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")

## 1.3 View the First Five Rows

The first five records are displayed to gain an initial understanding of the dataset structure and available features.

In [5]:
print(df.head())

   customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract Pape

## Observation

The dataset loaded successfully. The first five rows provide an initial view of the customer records and confirm that all columns have been imported correctly.

## 1.4 Check Dataset Dimensions

The number of rows and columns is examined to understand the size and overall structure of the dataset.

In [6]:
print("Rows and Columns:", df.shape)

Rows and Columns: (7043, 21)


### Observation

The dataset contains **7,043 customer records** and **21 features**, providing sufficient data to analyze customer churn patterns. The dataset includes customer demographics, services, billing information, and churn status.

## 1.5 View Column Names

The column names are reviewed to understand the available features and verify that the dataset has been imported correctly.

In [7]:
print(df.columns)

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='str')


## Observation

The dataset contains customer demographic information, subscribed services, billing details, contract information, and the target variable (`Churn`). These features will help identify factors influencing customer churn.

## 1.6 Dataset Information

The dataset information is examined to review data types, non-null values, and overall data quality.

In [8]:
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

## Observation

The dataset contains 7,043 records with a mix of categorical and numerical features. The `TotalCharges` column is stored as an object instead of a numeric data type, indicating that data cleaning is required before analysis.

## 1.7 Summary Statistics for Numerical Columns

Descriptive statistics are generated for numerical columns to understand their distribution and identify potential anomalies.

In [9]:
print(df.describe())

       SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000


## Observation

The summary statistics provide insights into customer tenure, monthly charges, and senior citizen status. These metrics help understand the distribution and variability of numerical features.

## 1.8 Summary Statistics for Categorical Columns

Summary statistics for categorical columns are generated to examine the frequency and distribution of customer attributes.

In [10]:
print(df.describe(include="object"))

        customerID gender Partner Dependents PhoneService MultipleLines  \
count         7043   7043    7043       7043         7043          7043   
unique        7043      2       2          2            2             3   
top     7590-VHVEG   Male      No         No          Yes            No   
freq             1   3555    3641       4933         6361          3390   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7043           7043         7043             7043   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3498         3088             3095   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7043        7043            7043            7043   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq

C:\Users\Admin\AppData\Local\Temp\ipykernel_12072\1129007761.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print(df.describe(include="object"))


## Observation

The categorical summary displays the number of unique values, the most frequent category, and its frequency. This helps understand the distribution of customer attributes such as contract type, internet service, and payment method.

## 1.9 Check for Missing Values

The dataset is checked for missing values to identify columns that may require cleaning before analysis.

In [11]:
print(df.isnull().sum())

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


## Observation

No missing values are detected using `isnull()`. However, some columns, particularly `TotalCharges`, may contain blank strings instead of true null values. These will be handled during the data cleaning stage.

## 1.10 Check for Duplicate Records

The dataset is examined for duplicate records to ensure each customer is represented only once.

In [12]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


## Observation

No duplicate customer records were found, indicating that each row represents a unique customer.

## 1.11 Count Unique Values in Each Column

The number of unique values in each column is reviewed to better understand the variability of each feature and identify identifier columns.

In [13]:
print(df.nunique())

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64


## Observation

The unique value counts confirm that `customerID` is a unique identifier, while other columns contain categorical or numerical values suitable for segmentation and analysis.

## 1.12 Explore Categorical Variables

The categorical variables are explored to understand customer demographics, subscribed services, contract types, payment methods, and churn categories.

In [14]:
for col in df.select_dtypes(include="object"):
    print(f"\n----- {col} -----")
    print(df[col].value_counts())


----- customerID -----
customerID
7590-VHVEG    1
5575-GNVDE    1
3668-QPYBK    1
7795-CFOCW    1
9237-HQITU    1
             ..
6840-RESVB    1
2234-XADUH    1
4801-JZAZL    1
8361-LTMKD    1
3186-AJIEK    1
Name: count, Length: 7043, dtype: int64

----- gender -----
gender
Male      3555
Female    3488
Name: count, dtype: int64

----- Partner -----
Partner
No     3641
Yes    3402
Name: count, dtype: int64

----- Dependents -----
Dependents
No     4933
Yes    2110
Name: count, dtype: int64

----- PhoneService -----
PhoneService
Yes    6361
No      682
Name: count, dtype: int64

----- MultipleLines -----
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64

----- InternetService -----
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

----- OnlineSecurity -----
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, 

C:\Users\Admin\AppData\Local\Temp\ipykernel_12072\2779499162.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in df.select_dtypes(include="object"):


## Observation

The categorical variables contain customer service preferences, contract types, payment methods, and churn status. Reviewing these categories helps identify data consistency issues and prepares the dataset for exploratory analysis.

# Part 1 Summary

### Key Findings

- The dataset contains **7,043 customer records** and **21 columns**.
- No duplicate records were identified.
- Most features are categorical and describe customer demographics, subscribed services, billing, and contract information.
- The `customerID` column uniquely identifies each customer.
- The `Churn` column is the target variable for analysis.
- The `TotalCharges` column is stored as an object and requires conversion to a numeric data type.
- Although no missing values are reported by `isnull()`, blank values in `TotalCharges` will need to be addressed during data cleaning.

# Part 2: Data Cleaning and Preparation

## Objective

The objective of this stage is to clean and prepare the Telco Customer Churn dataset for exploratory data analysis by correcting data types, handling missing values, and exporting a clean dataset for further analysis.

## 2.1 Convert TotalCharges to Numeric

The **TotalCharges** column is stored as a text data type due to blank entries. It is converted to a numeric data type to enable calculations and statistical analysis.

In [17]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

print(df["TotalCharges"].dtype)

float64


### Observation

The **TotalCharges** column has been successfully converted to a numeric data type. Any blank or invalid values have been converted to missing values for further cleaning.

## 2.2 Check Missing Values After Conversion

After converting the **TotalCharges** column, the dataset is checked again to identify any missing values created during the conversion process.

In [19]:
print(df.isnull().sum())

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64


### Observation

The missing values identified in the **TotalCharges** column correspond to records where the original values were blank.

## 2.3 Remove Missing Records

Rows containing missing values are removed because they represent only a small portion of the dataset and are unlikely to impact the overall analysis.

In [20]:
df = df.dropna()

print(df.shape)

(7032, 21)


### Observation

Rows containing missing values have been removed, resulting in a clean dataset for further analysis.

## 2.4 Verify the Cleaned Dataset

The cleaned dataset is verified to ensure that all missing values have been handled and the data types are appropriate for analysis.

In [21]:
print(df.info())

<class 'pandas.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   str    
 1   gender            7032 non-null   str    
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   str    
 4   Dependents        7032 non-null   str    
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   str    
 7   MultipleLines     7032 non-null   str    
 8   InternetService   7032 non-null   str    
 9   OnlineSecurity    7032 non-null   str    
 10  OnlineBackup      7032 non-null   str    
 11  DeviceProtection  7032 non-null   str    
 12  TechSupport       7032 non-null   str    
 13  StreamingTV       7032 non-null   str    
 14  StreamingMovies   7032 non-null   str    
 15  Contract          7032 non-null   str    
 16  PaperlessBilling  7032 non-null   str    
 17  PaymentMeth

### Observation

The dataset has been successfully cleaned and verified. All columns now have appropriate data types, and no missing values remain.

## 2.5 Export the Cleaned Dataset

The cleaned dataset is exported as a CSV file for use in exploratory data analysis and Power BI dashboard development.

In [22]:
df.to_csv("Telco_Customer_Churn_Cleaned.csv", index=False)

print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


### Observation

The cleaned dataset has been successfully exported and is ready for the next stage of the project.

# Part 2 Summary

### Key Cleaning Steps

- The **TotalCharges** column was successfully converted from a text (`object`) data type to a numeric data type.
- Blank values in the **TotalCharges** column were converted to missing values during the conversion process.
- A total of **11 records** containing missing values were identified after the conversion.
- The missing records were removed because they represented only a small proportion of the dataset and would have minimal impact on the overall analysis.
- The cleaned dataset was verified to ensure that all columns have appropriate data types and that no missing values remain.
- The cleaned dataset was exported as **`Telco_Customer_Churn_Cleaned.csv`** for use in exploratory data analysis and Power BI dashboard development.

# Part 3: Exploratory Data Analysis

## Objective

The objective of this stage is to analyze the cleaned Telco Customer Churn dataset to identify customer behavior patterns, measure churn across different customer segments, and generate actionable business insights that support data-driven decision-making.

## 3.1 Overall Customer Distribution

The distribution of retained and churned customers is examined to understand the overall composition of the customer base before performing detailed churn analysis.

In [34]:
print("Total Customers:", len(df))

print("\nCustomer Distribution:")

print(df["Churn"].value_counts())

Total Customers: 7032

Customer Distribution:
Churn
No     5163
Yes    1869
Name: count, dtype: int64


### Business Insight

The cleaned dataset contains **7,032 customers**, of which **5,163 customers (73.4%)** were retained and **1,869 customers (26.6%)** had churned. The dataset is moderately imbalanced, with retained customers forming the majority, which is an important consideration when interpreting churn patterns.

## 3.2 Calculate Overall Churn Rate

The overall churn rate is calculated to measure the percentage of customers who discontinued the service. This serves as a key business metric for evaluating customer retention.

In [24]:
churn_rate = (df["Churn"] == "Yes").mean() * 100

print(f"Overall Churn Rate: {churn_rate:.2f}%")

Overall Churn Rate: 26.58%


### Business Insight

The overall customer churn rate is **26.58%**, indicating that approximately **1 in every 4 customers** discontinued the service. This suggests a relatively high customer attrition rate and highlights the need to identify the key factors contributing to customer churn.

## 3.3 Churn Rate by Gender

The churn rate is calculated for male and female customers to determine whether gender has a meaningful relationship with customer retention.

In [35]:
gender_churn_rate = (
    pd.crosstab(df["gender"], df["Churn"], normalize="index") * 100
).round(2)

print(gender_churn_rate)

Churn      No    Yes
gender              
Female  73.04  26.96
Male    73.80  26.20


### Business Insight

Female customers have a slightly higher churn rate (**26.96%**) than male customers (**26.20%**). However, the difference is less than **1 percentage point**, indicating that **gender has minimal influence on customer churn** and is unlikely to be a primary driver of customer attrition.

## 3.4 Churn Rate by Senior Citizen Status

The churn rate is calculated for senior citizens and non-senior citizens to determine whether age group influences customer retention.

In [37]:
senior_churn_rate = (
    pd.crosstab(df["SeniorCitizen"], df["Churn"], normalize="index") * 100
).round(2)

senior_churn_rate.index = ["Non-Senior", "Senior Citizen"]

print(senior_churn_rate)

Churn              No    Yes
Non-Senior      76.35  23.65
Senior Citizen  58.32  41.68


### Business Insight

Senior citizens have a significantly higher churn rate (**41.68%**) compared to non-senior customers (**23.65%**). This suggests that senior citizens are considerably more likely to discontinue the service and may require targeted retention strategies, such as personalized customer support, loyalty programs, or service plans tailored to their needs.

## 3.5 Churn Rate by Contract Type

The churn rate is calculated for each contract type to determine how customer commitment influences customer retention.

In [38]:
contract_churn_rate = (
    pd.crosstab(df["Contract"], df["Churn"], normalize="index") * 100
).round(2)

print(contract_churn_rate)

Churn              No    Yes
Contract                    
Month-to-month  57.29  42.71
One year        88.72  11.28
Two year        97.15   2.85


### Business Insight

Customers with **month-to-month contracts** have the highest churn rate (**42.71%**), whereas customers with **one-year contracts** have a much lower churn rate (**11.28%**). Customers with **two-year contracts** exhibit the lowest churn rate (**2.85%**). This indicates that **long-term contracts are strongly associated with higher customer retention**, making contract commitment one of the most influential factors affecting customer churn.

## 3.6 Churn Rate by Payment Method

The churn rate is calculated for each payment method to determine whether customers using different payment options exhibit different retention patterns.

In [40]:
payment_churn_rate = (
    pd.crosstab(df["PaymentMethod"], df["Churn"], normalize="index") * 100
).round(2)

print(payment_churn_rate)

Churn                         No    Yes
PaymentMethod                          
Bank transfer (automatic)  83.27  16.73
Credit card (automatic)    84.75  15.25
Electronic check           54.71  45.29
Mailed check               80.80  19.20


### Business Insight

Customers using **Electronic Check** have the highest churn rate (**45.29%**), which is significantly higher than all other payment methods. In contrast, customers using **Credit Card (automatic)** (**15.25%**) and **Bank Transfer (automatic)** (**16.73%**) have the lowest churn rates. This suggests that customers enrolled in **automatic payment methods are more likely to remain with the company**, while Electronic Check users represent a high-risk customer segment requiring targeted retention efforts.

## 3.7 Churn Rate by Internet Service

The churn rate is calculated for each internet service type to determine whether the type of internet service influences customer retention.

In [41]:
internet_churn_rate = (
    pd.crosstab(df["InternetService"], df["Churn"], normalize="index") * 100
).round(2)

print(internet_churn_rate)

Churn               No    Yes
InternetService              
DSL              81.00  19.00
Fiber optic      58.11  41.89
No               92.57   7.43


### Business Insight

Customers using **Fiber Optic** internet service have the highest churn rate (**41.89%**), which is more than double the churn rate of **DSL** customers (**19.00%**) and significantly higher than customers with **no internet service** (**7.43%**). This suggests that customers subscribed to Fiber Optic services are at a much greater risk of churning, indicating a need to investigate factors such as pricing, service quality, or customer satisfaction within this segment.

## 3.8 Churn Rate by Tenure Group

Customers are grouped based on their tenure to examine how the length of their relationship with the company influences customer churn.

In [42]:
df["TenureGroup"] = pd.cut(
    df["tenure"],
    bins=[0, 12, 24, 48, 72],
    labels=["0-12 Months", "13-24 Months", "25-48 Months", "49-72 Months"],
    include_lowest=True
)

tenure_churn_rate = (
    pd.crosstab(df["TenureGroup"], df["Churn"], normalize="index") * 100
).round(2)

print(tenure_churn_rate)

Churn            No    Yes
TenureGroup               
0-12 Months   52.32  47.68
13-24 Months  71.29  28.71
25-48 Months  79.61  20.39
49-72 Months  90.49   9.51


### Business Insight

Customers in the **0–12 months** tenure group have the highest churn rate (**47.68%**), indicating that nearly half of new customers discontinue the service within their first year. The churn rate steadily decreases as customer tenure increases, reaching only **9.51%** for customers with **49–72 months** of tenure. This suggests that customer loyalty strengthens over time, making the first year a critical period for customer retention initiatives.

## 3.9 Average Monthly Charges by Churn Status

The average monthly charges are compared between retained and churned customers to determine whether pricing is associated with customer churn.

In [43]:
monthly_charge = (
    df.groupby("Churn")["MonthlyCharges"]
      .mean()
      .round(2)
)

print(monthly_charge)

Churn
No     61.31
Yes    74.44
Name: MonthlyCharges, dtype: float64


### Business Insight

Churned customers have a higher average monthly charge (**74.44**) than retained customers (**61.31**). This suggests that customers with higher monthly charges are more likely to discontinue the service, indicating that pricing or the perceived value of premium plans may influence customer retention.

## 3.10 Correlation Analysis

The relationship between numerical variables is examined to understand how customer tenure, monthly charges, and total charges are associated with one another.

In [44]:
correlation_matrix = df[["tenure", "MonthlyCharges", "TotalCharges"]].corr()

print(correlation_matrix.round(2))

                tenure  MonthlyCharges  TotalCharges
tenure            1.00            0.25          0.83
MonthlyCharges    0.25            1.00          0.65
TotalCharges      0.83            0.65          1.00


### Business Insight

A **strong positive correlation** exists between **tenure** and **TotalCharges** (**0.83**), indicating that customers who remain with the company longer tend to accumulate higher total charges. **MonthlyCharges** also show a **moderate positive correlation** with **TotalCharges** (**0.65**), while the relationship between **MonthlyCharges** and **tenure** is relatively weak (**0.25**). This suggests that total customer value is driven primarily by customer retention rather than higher monthly pricing alone.

# Part 3 Summary

### Key Findings

- The overall customer churn rate is **26.58%**, indicating that approximately **1 in every 4 customers** discontinued the service.
- Gender has minimal influence on customer churn, with only a small difference in churn rates between male and female customers.
- Senior citizens have a significantly higher churn rate than non-senior customers, making them a higher-risk customer segment.
- Customers with **month-to-month contracts** exhibit the highest churn rate, while **two-year contracts** have the lowest churn rate, highlighting contract duration as a major factor influencing customer retention.
- Customers using **Electronic Check** have the highest churn rate, whereas customers using automatic payment methods experience substantially lower churn.
- **Fiber Optic** customers have the highest churn rate among internet service types, indicating the need for further investigation into service quality, pricing, or customer satisfaction.
- Customers within their **first year of service** are the most likely to churn, while long-term customers demonstrate significantly higher retention.
- Churned customers have higher average monthly charges than retained customers, suggesting that pricing or the perceived value of premium plans may influence customer decisions.
- Correlation analysis indicates that customer lifetime value is strongly associated with tenure, emphasizing the importance of long-term customer retention strategies.

# Part 4: Feature Engineering and Data Preparation

## Objective

The objective of this stage is to create additional analytical features that improve reporting and visualization in Power BI. These derived columns simplify business analysis and support the development of interactive dashboards.

## 4.1 Create Customer Status

A customer status column is created to provide more meaningful labels for reporting and dashboard visualizations.

In [45]:
df["CustomerStatus"] = df["Churn"].replace({
    "Yes": "Churned",
    "No": "Retained"
})

print(df[["Churn", "CustomerStatus"]].head())

  Churn CustomerStatus
0    No       Retained
1    No       Retained
2   Yes        Churned
3    No       Retained
4   Yes        Churned


### Observation

A new **CustomerStatus** column was successfully created by replacing the original **Churn** values with more descriptive labels (**Retained** and **Churned**). This improves readability and makes dashboard reporting easier to understand.

## 4.2 Create Tenure Category

Customers are grouped into tenure categories to simplify retention analysis across different stages of the customer lifecycle.

In [50]:
df["TenureCategory"] = pd.cut(
    df["tenure"],
    bins=[0, 12, 24, 48, 72],
    labels=[
        "0-12 Months",
        "13-24 Months",
        "25-48 Months",
        "49-72 Months"
    ],
    include_lowest=True
)

print(df["TenureCategory"].value_counts())

TenureCategory
49-72 Months    2239
0-12 Months     2175
25-48 Months    1594
13-24 Months    1024
Name: count, dtype: int64


### Observation

The **TenureCategory** column was successfully created by grouping customers into four tenure ranges based on their length of service. This derived feature simplifies customer lifecycle analysis and supports more effective reporting and visualization in Power BI.

## 4.3 Create Monthly Charge Category

Monthly charges are grouped into pricing bands to simplify customer pricing analysis.

In [49]:
df["MonthlyChargeCategory"] = pd.cut(
    df["MonthlyCharges"],
    bins=[0, 35, 70, 120],
    labels=[
        "Low",
        "Medium",
        "High"
    ]
)

print(df["MonthlyChargeCategory"].value_counts())

MonthlyChargeCategory
High      3581
Low       1729
Medium    1722
Name: count, dtype: int64


### Observation

The **MonthlyChargeCategory** column was successfully created by grouping customers into **Low**, **Medium**, and **High** monthly charge bands. This derived feature simplifies pricing analysis and supports clearer segmentation and visualization in the Power BI dashboard.

## 4.4 Calculate Revenue at Risk

Revenue at Risk is calculated by summing the monthly charges of churned customers. This metric estimates the recurring monthly revenue lost due to customer churn.

In [54]:
revenue_at_risk = df.loc[
    df["Churn"] == "Yes",
    "MonthlyCharges"
].sum()

print(f"Monthly Revenue at Risk: {revenue_at_risk:,.2f}")

Monthly Revenue at Risk: 139,130.85


### Observation

The **Monthly Revenue at Risk** was successfully calculated as **139130.85**, representing the estimated recurring monthly revenue associated with customers who have churned. This metric can be used as a key performance indicator (KPI) in the Power BI dashboard to monitor the financial impact of customer churn.

## 4.5 Verify Newly Created Features

The newly created analytical columns are reviewed to ensure they were generated correctly before exporting the final dataset.

In [51]:
print(df.head())

   customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ...        Contract  \
0  No phone service             DSL             No  ...  Month-to-month   
1                No             DSL            Yes  ...        One year   
2                No             DSL            Yes  ...  Month-to-month   
3  No phone service             DSL            Yes  ...        One year   
4                No     Fiber optic             No  ...  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges TotalCh

### Observation

The newly created features (**CustomerStatus**, **TenureCategory**, and **MonthlyChargeCategory**) were successfully verified in the dataset. The preview confirms that the derived columns have been added correctly and are ready to support analysis and visualization in Power BI.

## 4.6 Export Final Dataset

The final prepared dataset is exported for use in Power BI dashboard development.

In [53]:
df.to_csv("Telco_Customer_Churn_Cleaned_Final.csv", index=False)

print("Final dataset exported successfully.")

Final dataset exported successfully.


### Observation

The final prepared dataset was successfully exported as **Telco_Customer_Churn_Cleaned_Final.csv**. It contains the cleaned data and newly engineered features, making it ready for Power BI dashboard development and business reporting.

# Part 4 Summary

### Key Enhancements

- A new **CustomerStatus** column was created to replace the original **Churn** values with more descriptive labels (**Retained** and **Churned**) for improved reporting.
- Customers were grouped into **TenureCategory** bands to simplify customer lifecycle and retention analysis.
- Monthly charges were categorized into **Low**, **Medium**, and **High** pricing bands through the **MonthlyChargeCategory** feature.
- The **Monthly Revenue at Risk** metric was calculated as **139130.85**, providing an estimate of the recurring monthly revenue associated with churned customers.
- All newly engineered features were successfully verified and confirmed to be correctly added to the dataset.
- The final dataset was exported as **Telco_Customer_Churn_Final.csv**, making it ready for Power BI dashboard development and interactive business reporting.

# Part 5: Power BI Dashboard Development

## Objective

The objective of this stage is to transform the prepared dataset into an interactive Power BI dashboard that enables stakeholders to monitor customer churn, identify key business drivers, and support data-driven decision-making.

## 5.1 Import the Final Dataset

The final prepared dataset is imported into Power BI Desktop to begin dashboard development.

## 5.2 Create Key Performance Indicators (KPIs)

Key performance indicators are created to provide a high-level overview of customer churn and business performance.

## 5.3 Build Interactive Visualizations

Interactive visualizations are created to analyze customer churn across different customer segments and business dimensions.

## 5.4 Dashboard Design

The dashboard is designed using a clean and interactive layout to improve readability and support effective business decision-making.

## 5.5 Business Recommendations

Based on the analysis, several recommendations are proposed to improve customer retention and reduce churn.

# Part 5 Summary

### Key Outcomes

- Imported the prepared dataset into Power BI.
- Developed an interactive dashboard using key business metrics and visualizations.
- Highlighted the major drivers of customer churn through interactive analysis.
- Presented actionable business recommendations based on the findings.
- Delivered a complete end-to-end customer churn analysis project, from data inspection and cleaning to business reporting and visualization.